### ISSUE TITLE
Implement De-Batching

### ISSUE THREAD

**Issue by gjedlicska** [specklesystems/speckle-server#1077]
Example commit:
https://speckle.xyz/streams/c239718aff/commits/b3a8cfb97d
OS:
Linux
Browser:
Chrome and Firefox

**Comment by AlexandruPopovici** [comment]
I edited the title, because SAO has nothing to do with the issue. I get the same performance issues without SAO on chromium on windows. However everything runs fine on Firefox

**Comment by AlexandruPopovici** [comment]
Right, so after investigating this issue it looks like there are multiple problems to why for the mentioned stream performs so badly. 

First one is related to this [issue ](https://github.com/specklesystems/speckle-server/issues/1051). Currently, the viewer batches together indiscriminately everything it can. This can lead to extremely large VBOs which we need to bind to at runtime. For this particular stream, however, it's not the batching's fault since there is nothing to batch. There is only a single mesh of around 1.5 million triangles. Even so, the result is the same and consequences are the same.

Second one is more complicated. Before going into it, here's an [epic good reference](https://fgiesen.wordpress.com/2011/07/09/a-trip-through-the-graphics-pipeline-2011-index/) regarding what we'll be discussing. We'll be using the stream mentioned in the issue description going forward since it's quite an extreme case, but relevant nonetheless. 
The stream holds a single mesh. It's a relatively large one, but not large enough to cause such performance issues. The viewer can easily handle double that amount of triangles with no visible performance drop, so something else must be going on. I tried looking at how the mesh's index buffer is built by rendering sequential parts of it. Here is for example the first 1/10 of the triangles:

![tris`](https://user-images.githubusercontent.com/11405391/194364046-4408f288-e4da-4da6-8e81-a95517a52787.PNG)

And now a close-up:

![tris2](https://user-images.githubusercontent.com/11405391/194364165-7b47be6d-023d-4741-8218-8cfcb7f517a7.PNG)

As you can image, the following 1/10's of the index buffer continue in the same fashion. This randomization in locality is causing the strange performance issues. Whenever vertices are being pulled on the GPU for processing they are first looked for in the vertex cache (which is not large, like any hw cache). A large number of such cache misses have a negative impact of performance. Looking at the mesh topology one can safely assume that there are going to be a lot of cache misses. A ridiculous amount probably.
In order to put this theory to test I got the model into Blender and spatially sorted the triangles. Now if we render the first 1/10 of the index buffer we get:

![tris3](https://user-images.githubusercontent.com/11405391/194367323-b603894e-cd40-40c4-afa5-82f710fe478a.PNG)

Which is spatially the first 1/10 of the mesh. Let's look at the performance comparison between the sorted and non-sorted models:
- Original, non-sorted model -> **26 FPS**
- Sorted model -> **83 FPS**

In practice I don't think we can solve these types of issues by reordering the triangles at runtime. I don't think we have the time budget to do that, even if there are algorithms specifically designed for it, like [this one](https://tomforsyth1000.github.io/papers/fast_vert_cache_opt.html). Instead what we would need to do is to split these large meshes into multiple submeshes. Basically, the opposite of what we're doing now with batching. So I guess we'll have to do both batching and de-batching inside the viewer :). De-batching should also help with the first issue mentioned above and get rid of huge VBOs.

In order to get the best performance in situations like this one, it's not enough to solve just one of the problems described above. Both of them need to be handled in order to satisfy the peculiarities of most GPUs. For example, with the sorted model, my laptop and phone GPU showed a huge performance increase. Gergo's laptop and phone as well. Dim's Apple laptop on the other hand showed little to no improvement (but his Apple phone did if I remember correctly). Bilal's integrated GPU also showed little performance increase with the sorted model. That's why I stress the fact that different GPUs have different bottlenecks for this particular stream, so in order to satisfy all of them, all the probable causes for bottlenecks need to be handled. One more detail regarding the mentioned testing and debugging, namely all of it has been done on various *low-range* GPUs. All GPUs we tested on are either integrated GPUs, either phone GPUs (which sometimes are faster than CPU integrated ones). One would not have such issues on any regular GPU, mobile or otherwise.

**Comment by gjedlicska** [comment]
Thanks for the detailed analysis
